pkg/buildscript/buildscript.go

```diff
@@ -0,0 +1,179 @@
+		return nil, errs.Wrap(err, "Could not unmarshal buildscript")
+	}
+
+	be, err := raw.MarshalBuildExpression()
+	if err != nil {
+		return nil, errs.Wrap(err, "Could not marshal build expression from raw")
+	}
+
+	expr, err := buildexpression.Unmarshal(be)
+	if err != nil {
+		return nil, errs.Wrap(err, "Could not unmarshal build expression")
+	}
+
+	return &BuildScript{expr, raw.AtTime}, nil
+}
+
+// UnmarshalBuildExpression will create buildscript using an existing build expression
+func UnmarshalBuildExpression(b []byte, atTime *time.Time) (*BuildScript, error) {
+	expr, err := buildexpression.Unmarshal(b)
+	if err != nil {
+		return nil, errs.Wrap(err, "Could not parse build expression")
+	}
+
+	return unmarshalBuildExpressionTyped(expr, atTime)
+}
+
+func unmarshalBuildExpressionTyped(expr *buildexpression.BuildExpression, atTime *time.Time) (*BuildScript, error) {
+	// Copy incoming build expression to keep any modifications local.
+	var err error
+	expr, err = expr.Copy()
+	if err != nil {
+		return nil, errs.Wrap(err, "Could not copy build expression")
+	}
+
+	// Update old expressions that bake in at_time as a timestamp instead of as a variable.
+	err = expr.MaybeSetDefaultAtTime(atTime)
+	if err != nil {
+		return nil, errs.Wrap(err, "Could not set default timestamp")
+	}
+
+	return &BuildScript{expr, atTime}, nil
+}
+
+// MarshalBuildExpression translates our buildscript into a build expression
+// The actual logic for this lives under the MarshalJSON methods below, named that way because that's what the json
+// marshaller is expecting to find.
+func (b *BuildScript) MarshalBuildExpression() ([]byte, error) {
+	bytes, err := json.MarshalIndent(b.buildexpression, "", "  ")
+	if err != nil {
+		return nil, errs.Wrap(err, "Could not marshal build script to build expression")
+	}
+	return bytes, nil
+}
+
+// Requirements returns the requirements in the Buildscript
+// It returns an error if the requirements are not found or if they are malformed.
+func (b *BuildScript) Requirements() ([]types.Requirement, error) {
+	return b.buildexpression.Requirements()
+}
+
+// RequirementNotFoundError aliases the buildexpression error type, which can otherwise not be checked as its internal
+type RequirementNotFoundError = buildexpression.RequirementNotFoundError
+
+// UpdateRequirement updates the Buildscripts requirements based on the operation and requirement.
+func (b *BuildScript) UpdateRequirement(operation types.Operation, requirement types.Requirement) error {
+	return b.buildexpression.UpdateRequirement(operation, requirement)
+}
+
+func (b *BuildScript) UpdatePlatform(operation types.Operation, platformID strfmt.UUID) error {
+	return b.buildexpression.UpdatePlatform(operation, platformID)
+}
+
+func (b *BuildScript) AtTime() *time.Time {
+	return b.atTime
+}
+
+func (b *BuildScript) SetAtTime(t time.Time) {
+	b.atTime = &t
+}
+
+func (b *BuildScript) SetDefaultAtTime() error {
+	return b.buildexpression.SetDefaultAtTime()
```

Comment:
Do we need to set `b.atTime = nil` here, too?

Reply:
I've renamed this function. It wasn't really setting any atTime value, it was merely forcing it to be set to `$at_time`, which isn't appropriate on the buildscript level.